SteamKit2/SteamKit2/Util/HardwareUtils.cs

```diff
@@ -0,0 +1,398 @@
+                  WHERE Index = {0}", index
+                ).FirstOrDefault();
+
+            if ( bootableDisk == null )
+            {
+                return base.GetDiskId();
+            }
+
+            string serialNumber = (string)bootableDisk["SerialNumber"];
+
+            return Encoding.UTF8.GetBytes( serialNumber );
+        }
+
+
+        IEnumerable<ManagementObject> WmiQuery( string queryFormat, params object[] args )
+        {
+            string query = string.Format( queryFormat, args );
+
+            var searcher = new ManagementObjectSearcher( query );
+
+            return searcher.Get().Cast<ManagementObject>();
+        }
+    }
+
+    class LinuxInfoProvider : DefaultInfoProvider
+    {
+        public override byte[] GetMachineGuid()
+        {
+            string[] machineFiles =
+            {
+                "/etc/machine-id", // present on at least some gentoo systems
+                "/var/lib/dbus/machine-id",
+                "/sys/class/net/eth0/address",
+                "/sys/class/net/eth1/address",
+                "/sys/class/net/eth2/address",
+                "/sys/class/net/eth3/address",
+                "/etc/hostname",
+            };
+
+            foreach ( var fileName in machineFiles )
+            {
+                try
+                {
+                    return File.ReadAllBytes( fileName );
+                }
+                catch
+                {
+                    // if we can't read a file, continue to the next until we hit one we can
+                    continue;
+                }
+            }
+
+            return base.GetMachineGuid();
+        }
+
+        public override byte[] GetDiskId()
+        {
+            string[] bootParams = GetBootOptions();
+
+            string[] paramsToCheck =
+            {
+                "root=UUID=",
+                "root=PARTUUID=",
+            };
+
+            foreach ( string param in paramsToCheck )
+            {
+                string paramValue = GetParamValue( bootParams, param );
+
+                if ( !string.IsNullOrEmpty( paramValue ) )
+                {
+                    return Encoding.UTF8.GetBytes( paramValue );
+                }
+            }
+
+            string[] diskUuids = GetDiskUUIDs();
+
+            if ( diskUuids.Length > 0 )
+            {
+                return Encoding.UTF8.GetBytes( diskUuids.FirstOrDefault() );
+            }
+
+            return base.GetDiskId();
+        }
+
+
+        string[] GetBootOptions()
+        {
+            string bootOptions;
+
+            try
+            {
+                bootOptions = File.ReadAllText( "/proc/cmdline" );
+            }
+            catch
+            {
+                return new string[0];
+            }
+
+            return bootOptions.Split( ' ' );
+        }
+        string[] GetDiskUUIDs()
+        {
+            string[] files;
+
+            try
+            {
+                files = Directory.GetFiles( "/dev/disk/by-uuid" );
```

Comment:
Does this return files in any order? If so, could this change just by adding or removing a USB flash drive, for example?